keras/optimizers/base_optimizer.py

```diff
@@ -304,19 +327,100 @@ def apply(self, grads, trainable_variables=None):
+        """
+        if self.gradient_accumulation_steps:
+            is_update_step = (
+                self.iterations + 1
+            ) % self.gradient_accumulation_steps == 0
+
+            def _update_step_fn(self, grads, trainable_variables):
+                # Run update step with accumulated grads + reset accumulators
+                steps = self.gradient_accumulation_steps
+                grads = [
+                    (grads[i] + self._accumulated_gradients[i]) / steps
+                    for i in range(len(grads))
+                ]
+                self._backend_update_step(
+                    grads, trainable_variables, self.learning_rate
+                )
+                self._backend_reset_gradient_accumulators()
+
+            def _grad_accumulation_fn(self, grads):
+                # Update gradient accumulators
+                self._backend_increment_gradient_accumulators(grads)
+
+            ops.cond(
+                is_update_step,
+                lambda: _update_step_fn(self, grads, trainable_variables),
+                lambda: _grad_accumulation_fn(self, grads),
+            )
+        else:
+            # Run udpate step.
+            self._backend_update_step(
+                grads, trainable_variables, self.learning_rate
+            )
+
+        if self.use_ema:
+            self._update_model_variables_moving_average(
+                self._trainable_variables
+            )
+            if self.ema_overwrite_frequency:
+                # Only when self.ema_overwrite_frequency is not None, we
+                # overwrite the model variables.
+                should_overwrite_model_vars = (
+                    self.iterations + 1
+                ) % self.ema_overwrite_frequency == 0
+                ops.cond(
+                    should_overwrite_model_vars,
+                    lambda: self._overwrite_model_variables_with_average_value(
+                        self._trainable_variables
+                    ),
+                    lambda: None,
+                )
+        # Update iteration counter.
+        self.iterations.assign_add(1)
+
+    def _backend_update_step(self, grads, trainable_variables, learning_rate):
+        """Collective update_step that can be overridden by the backend.
+
+        It is overridden by torch for performance reasons, and
+        by TF to support tf.distribute.
+        """
+        for grad, var in zip(grads, trainable_variables):
+            self.update_step(grad, var, learning_rate)
+
+    def _backend_reset_gradient_accumulators(self):
+        for g_acc in self._accumulated_gradients:
+            g_acc.assign(np.zeros(g_acc.shape, dtype=g_acc.dtype))
+
+    def _backend_increment_gradient_accumulators(self, grads):
+        new_g_accs = [
+            (grads[i] + self._accumulated_gradients[i])
+            for i in range(len(grads))
+        ]
+        for n_g_acc, g_acc in zip(new_g_accs, self._accumulated_gradients):
+            g_acc.assign(n_g_acc)
```

Comment:
In the tf.distribute context, I think this probably should use https://www.tensorflow.org/api_docs/python/tf/distribute/StrategyExtended#update

Reply:
Ok, but what about `self.iterations`? We update with a simple `assign`. When is it ok to use assign and when is it not?